crates/sui-core/src/authority/authority_store.rs

```diff
@@ -522,10 +524,17 @@ impl AuthorityStore {
         let mut write_batch = self.perpetual_tables.objects.batch();
 
         // Insert object
+        let StoreObjectPair(store_object, indirect_object) = object.clone().into();
         write_batch = write_batch.insert_batch(
             &self.perpetual_tables.objects,
-            std::iter::once((ObjectKey::from(object_ref), object)),
+            std::iter::once((ObjectKey::from(object_ref), store_object)),
         )?;
+        if let Some(indirect_obj) = indirect_object {
+            write_batch = write_batch.insert_batch(
```

Comment:
Are we always inserting the indirect object in the second table? That seems like it would not give us any benefit over the current approach

Reply:
this particular function is used for genesis only.
But overall, concern is valid. Let's discuss.
For normal code path merge is used. 
During merge multiple values are getting collapsed into single one, but this is done on a read path and/or during compaction process. So db might contain [Merge_OP1, Merge_OP2, ....] each of them carrying the value before compaction kicks in. 
Should we consider doing read before write instead? If we see that value exists, we can increment reference count instead

Reply:
ok, it looks like proper fix requires transaction usage. And we can't enable it now with the current pruner(because it uses `delete_range` functionality). 
We can land PR as it is for now. It'll lead to described inefficiency. And then resolve it as a followup once we enable optimistic transactions for this rocksdb instance.

But just to talk it through, if we do read before write without enabling transactions:
*  concurrent writes should work fine. Even if 2 writes read that the value is empty, they both will initiate merge with full value, which will be resolved by merge operation
* the potential problem is only possible in following scenario:
    * write sees that specific value exists in storage. Decides to do partial merge instead of full merge
    * pruning job decrements ref count for given value and sets it to 0
    * compaction kicks in, collapses all merge operators, and removes entry from the db because ref count is currently 0
    * write operation finally performs partial merge and commits. But now original value is absent from db